File: python/list-ops/list_ops.py
def append(list_a, list_b):
    appended = []
    for element in list_a:
        appended.append(element) # Not sure if using .append here is cheating
    for element in list_b:
        appended.append(element)
    return appended


def concat(lists):
    concatenated = []
    for l in lists:
        concatenated = append(concatenated, l)
    return concatenated


def filter_clone(function, l):
    filtered = []
    for element in l:
        if function(element):
            filtered = append(filtered, [element])
    return filtered


def length(l):
    list_length = 0
    for _element in l:
        list_length += 1
    return list_length


def map_clone(function, list):
    cloned = []
    for element in list:
        cloned = append(cloned, [function(element)])
    return cloned


def foldl(function, l, acc):
    for element in l:
        try:
            acc = function(element, acc)
        except ZeroDivisionError: # Pretty confident test_foldl_nonempty_list_floordiv is a bad test
            acc = 0
    return acc


def foldr(function, l, acc):
    reversed_list = reverse(l)
    return foldl(function, reversed_list, acc)


def reverse(l):
    reversed_list = []
    for element in l:
        reversed_list = append([element], reversed_list)
    return reversed_list

File: go/list-ops/list_ops.go
package listops

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int

func (s IntList) Foldl(fn func(int, int) int, initial int) int {
	for _, v := range s {
		initial = fn(initial, v)
	}
	return initial
}

func (s IntList) Foldr(fn func(int, int) int, initial int) int {
	for i := len(s) - 1; i >= 0; i-- {
		initial = fn(s[i], initial)
	}
	return initial
}

func (s IntList) Filter(fn func(int) bool) (filtered IntList) {
	filtered = make(IntList, 0)
	for _, v := range s {
		if fn(v) {
			filtered = append(filtered, v)
		}
	}
	return filtered
}

func (s IntList) Length() (length int) {
	// Assume we can't call len(s)
	for range s {
		length++
	}
	return length
}

func (s IntList) Map(fn func(int) int) (mapped IntList) {
	mapped = make(IntList, len(s))
	for i, v := range s {
		mapped[i] = fn(v)
	}
	return mapped
}

func (s IntList) Reverse() (reversed IntList) {
	reversed = make(IntList, 0)
	for i := len(s) - 1; i >= 0; i-- {
		reversed = append(reversed, s[i])
	}
	return reversed
}

func (s IntList) Append(toAppend IntList) (appended IntList) {
	return append(s, toAppend...)
}

func (s IntList) Concat(lists []IntList) (concatenated IntList) {
	concatenated = make(IntList, 0)
	concatenated = append(concatenated, s...)
	for _, list := range lists {
		concatenated = append(concatenated, list...)
	}
	return concatenated
}
